vertexai/genai/tokenizer/corpora_test.go

```diff
@@ -241,47 +242,66 @@ func TestCountTokensWithCorpora(t *testing.T) {
 	model := client.GenerativeModel(defaultModel)
 	ucr := newUdhrCorpus()
 
+	tok, err := New(defaultModel)
+	if err != nil {
+		log.Fatal(err)
+	}
+
 	corporaURL := "https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/packages/corpora/udhr.zip"
-	files, err := corporaGenerator(corporaURL)
+	corporaFiles, err := corporaGenerator(corporaURL)
 	if err != nil {
 		t.Fatalf("Failed to generate corpora: %v", err)
 	}
 
-	// Iterate over files generated by the generator function
-	for _, fileInfo := range files {
-		if ucr.shouldSkip(fileInfo.Name) {
-			fmt.Printf("Skipping file: %s\n", fileInfo.Name)
-			continue
-		}
+	// Create channels for work distribution and results collection
+	corporaChan := make(chan corporaInfo)
+	doneChan := make(chan bool)
 
-		enc, found := ucr.getEncoding(fileInfo.Name)
-		if !found {
-			fmt.Printf("No encoding found for file: %s\n", fileInfo.Name)
-			continue
-		}
+	worker := func() {
+		for corpora := range corporaChan {
+			if ucr.shouldSkip(corpora.Name) {
+				fmt.Printf("Skipping file: %s\n", corpora.Name)
+				continue
+			}
 
-		decodedContent, err := decodeBytes(enc, fileInfo.Content)
-		if err != nil {
-			log.Fatalf("Failed to decode bytes: %v", err)
-		}
+			enc, found := ucr.getEncoding(corpora.Name)
+			if !found {
+				fmt.Printf("No encoding found for file: %s\n", corpora.Name)
+				continue
+			}
 
-		tok, err := New(defaultModel)
-		if err != nil {
-			log.Fatal(err)
-		}
+			decodedContent, err := decodeBytes(enc, corpora.Content)
+			if err != nil {
+				log.Fatalf("Failed to decode bytes: %v", err)
+			}
 
-		localNtoks, err := tok.CountTokens(genai.Text(decodedContent))
-		if err != nil {
-			log.Fatal(err)
-		}
-		remoteNtoks, err := model.CountTokens(ctx, genai.Text(decodedContent))
-		if err != nil {
-			log.Fatal(fileInfo.Name, err)
-		}
-		if localNtoks.TotalTokens != remoteNtoks.TotalTokens {
-			t.Errorf("expected %d(remote count-token results), but got %d(local count-token results)", remoteNtoks, localNtoks)
+			localNtoks, err := tok.CountTokens(genai.Text(decodedContent))
+			if err != nil {
+				log.Fatal(err)
+			}
+			remoteNtoks, err := model.CountTokens(ctx, genai.Text(decodedContent))
+			if err != nil {
+				log.Fatal(corpora.Name, err)
+			}
+			if localNtoks.TotalTokens != remoteNtoks.TotalTokens {
+				t.Errorf("expected %d(remote count-token results), but got %d(local count-token results)", remoteNtoks, localNtoks)
+			}
 		}
+		doneChan <- true
```

Comment:
I think this code will work fine, but since your workers don't actually return any results, I recommend a slightly different pattern, using a `WorkGroup`.

This example shows how: https://gobyexample.com/waitgroups

Then you won't need the loop clearing the `doneChan` in the main goroutine; you'll just do a wait on the group.

Reply:
Done